### PR TITLE
add a conversation example

### PR DESCRIPTION
**Pull request by hubenchang0515**

### CODE REVIEW

**Comment by ldv-alt** [comment]
First of all, could you make sure it passes CI, please?

**Comment by hubenchang0515** [comment]
I fix the code on the basis of CI, but I seem to do a wrong operation (git push -f)
Is there any other mistake

**Comment by t8m** [comment]
You should add the tty_conv to the Makefile.am noinst_PROGRAMS.

File: examples/tty_conv.c [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+static void echoOnStdin()
+{
+    echoOff(fileno(stdin), 0);
+}
+
+/***************************************
+ * @brief read a line input
+ * @return the input string
+ * *************************************/
+static char* readline()
```
Comment (t8m) [inline examples/tty_conv.c:52]: A space should be before `*` not after.

File: examples/tty_conv.c [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+{
+    struct termio tty;
+    char input[PAM_MAX_RESP_SIZE];
+
+    flockfile(stdin);
+    int i = 0;
```
Comment (t8m) [inline examples/tty_conv.c:58]: Please avoid mixing declarations with statements.

File: examples/tty_conv.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ * @brief callback of PAM conversation
+ * @param[in] num_msg the count of message
+ * @param[in] msg PAM message
+ * @param[out] resp our response
+ * @param[in] appdata_ptr custom data passed by struct pam_conv.appdata_ptr
+ * @return state
+ * ************************************************/
+static int conversation(int num_msg, const struct pam_message** msg, struct pam_response **resp, void *appdata_ptr)
+{
+    // check the count of message
```
Comment (t8m) [inline examples/tty_conv.c:82]: Please avoid C++ style comments.

File: examples/tty_conv.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    // alloc memory for response
+    if ((resp[0] = malloc(num_msg * sizeof(struct pam_response))) == NULL)
+    {
+        fprintf(stderr, "bad alloc\n");
+        return PAM_BUF_ERR;
+    }
+
+    // response for message
+    for(int i = 0; i < num_msg; i++)
```
Comment (t8m) [inline examples/tty_conv.c:97]: Please move the declaration to the beginning of the block.

File: examples/tty_conv.c [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    return PAM_SUCCESS;
+}
+
+int main()
+{
+    atexit(echoOnStdin); // echo on while exist, like Ctrl+C on input password
+
+    struct pam_conv pam_conv = {conversation, NULL};
+    pam_handle_t *pamh;
```
Comment (t8m) [inline examples/tty_conv.c:138]: Please move the declarations.

**Review by t8m — CHANGES_REQUESTED** [review]

File: examples/tty_conv.c (hunk 2 above)
Comment (hubenchang0515) [reply examples/tty_conv.c:58]: Is it means I shoud write as this?
```c
struct termio tty;
char input[PAM_MAX_RESP_SIZE];
int i;

flockfile(stdin);
for (i = 0; i < PAM_MAX_RESP_SIZE; i++)
```

File: examples/tty_conv.c (hunk 3 above)
Comment (hubenchang0515) [reply examples/tty_conv.c:82]: what is C++ style comments? is it means i should write as this?
```c
/* check the count of message */
```

File: examples/tty_conv.c (hunk 2 above)
Comment (t8m) [reply examples/tty_conv.c:58]: yes

File: examples/tty_conv.c (hunk 3 above)
Comment (t8m) [reply examples/tty_conv.c:82]: yes

File: examples/tty_conv.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+static void echoOnStdin(void)
+{
+    echoOff(fileno(stdin), 0);
+}
+
+/***************************************
+ * @brief read a line input
+ * @return the input string
+ ***************************************/
+static char* readline(void)
```
Comment (t8m) [inline examples/tty_conv.c:52]: Nit: space before `*` not after.

File: examples/tty_conv.c [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+/**************************************************
+ * @brief callback of PAM conversation
+ * @param[in] num_msg the count of message
+ * @param[in] msg PAM message
+ * @param[out] resp our response
+ * @param[in] appdata_ptr custom data passed by struct pam_conv.appdata_ptr
+ * @return state
+ **************************************************/
+static int conversation(int num_msg, const struct pam_message** msg, struct pam_response **resp, void *appdata_ptr)
```
Comment (t8m) [inline examples/tty_conv.c:79]: Nit: space before `**` not after.

File: examples/tty_conv.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+    /* alloc memory for response */
+    if ((resp[0] = malloc(num_msg * sizeof(struct pam_response))) == NULL)
+    {
+        fprintf(stderr, "bad alloc\n");
+        return PAM_BUF_ERR;
+    }
+
+    /* response for message */
+    for(i = 0; i < num_msg; i++)
```
Comment (t8m) [inline examples/tty_conv.c:99]: Nit: space after `for`

File: examples/tty_conv.c [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        const struct pam_message* m = *msg + i;
+        struct pam_response* r = *resp + i;
```
Comment (t8m) [inline examples/tty_conv.c:102]: Nit: space before `*` not after.

File: examples/tty_conv.c [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    return PAM_SUCCESS;
+}
+
+int main(void)
+{
+    struct pam_conv pam_conv = {conversation, NULL};
+    pam_handle_t *pamh;
+
+    /* echo on while exist, like Ctrl+C on input password */
+    atexit(echoOnStdin);
```
Comment (t8m) [inline examples/tty_conv.c:141]: Please drop the trailing whitespace.

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few more nits

**Review by t8m — APPROVED** [review]

File: examples/tty_conv.c [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    if (off)
+    {
+        tty.c_lflag &= ~(ECHO | ECHOE | ECHOK | ECHONL);
+        (void) ioctl(fd, TCSETAF, &tty);
+    }
+    else
+    {
+        tty.c_lflag |= (ECHO | ECHOE | ECHOK | ECHONL);
+        (void) ioctl(fd, TCSETAW, &tty);
+    }
```
Comment (ldv-alt) [inline examples/tty_conv.c:29]: Why would you want to perform an ioctl TCSETAW operation on a descriptor if ioctl TCGETA failed?
A good example code shouldn't ignore the return value of ioctl calls, should it?

**Review by ldv-alt — CHANGES_REQUESTED** [review]

**Comment by hubenchang0515** [comment]
checked the return value of ioctl

**Review by ldv-alt — APPROVED** [review]
Assuming that all these commits are squashed, I'm OK with this PR.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged manually to master branch after squashing.

**Pull request closed without merging**